File: android-agent/src/test/java/io/opentelemetry/android/internal/features/networkattrs/CurrentNetworkAttributesExtractorTest.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.internal.features.networkattrs;

import static org.assertj.core.api.Assertions.entry;

import android.os.Build;
import io.opentelemetry.android.internal.services.network.data.Carrier;
import io.opentelemetry.android.internal.services.network.data.CurrentNetwork;
import io.opentelemetry.android.internal.services.network.data.NetworkState;
import io.opentelemetry.sdk.testing.assertj.OpenTelemetryAssertions;
import io.opentelemetry.semconv.SemanticAttributes;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
public class CurrentNetworkAttributesExtractorTest {

    final CurrentNetworkAttributesExtractor underTest = new CurrentNetworkAttributesExtractor();

    @Config(sdk = Build.VERSION_CODES.P)
    @Test
    public void getNetworkAttributes_withCarrier() {
        CurrentNetwork currentNetwork =
                CurrentNetwork.builder(NetworkState.TRANSPORT_CELLULAR)
                        .subType("aaa")
                        .carrier(
                                Carrier.builder()
                                        .id(206)
                                        .name("ShadyTel")
                                        .isoCountryCode("US")
                                        .mobileCountryCode("usa")
                                        .mobileNetworkCode("omg")
                                        .build())
                        .build();

        OpenTelemetryAssertions.assertThat(underTest.extract(currentNetwork))
                .containsOnly(
                        entry(SemanticAttributes.NETWORK_CONNECTION_TYPE, "cell"),
                        entry(SemanticAttributes.NETWORK_CONNECTION_SUBTYPE, "aaa"),
                        entry(SemanticAttributes.NETWORK_CARRIER_NAME, "ShadyTel"),
                        entry(SemanticAttributes.NETWORK_CARRIER_ICC, "US"),
                        entry(SemanticAttributes.NETWORK_CARRIER_MCC, "usa"),
                        entry(SemanticAttributes.NETWORK_CARRIER_MNC, "omg"));
    }

    @Config(sdk = Build.VERSION_CODES.O)
    @Test
    public void getNetworkAttributes_withoutCarrier() {
        CurrentNetwork currentNetwork =
                CurrentNetwork.builder(NetworkState.TRANSPORT_CELLULAR)
                        .subType("aaa")
                        .carrier(Carrier.builder().id(42).name("ShadyTel").build())
                        .build();

        OpenTelemetryAssertions.assertThat(underTest.extract(currentNetwork))
                .containsOnly(
                        entry(SemanticAttributes.NETWORK_CONNECTION_TYPE, "cell"),
                        entry(SemanticAttributes.NETWORK_CONNECTION_SUBTYPE, "aaa"));
    }
}

File: services-network/src/test/java/io/opentelemetry/android/internal/services/network/CurrentNetworkAttributesExtractorTest.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.internal.services.network

import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.kotlin.semconv.IncubatingApi
import io.opentelemetry.kotlin.semconv.NetworkAttributes
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

@OptIn(IncubatingApi::class)
class CurrentNetworkAttributesExtractorTest {
    private val underTest: CurrentNetworkAttributesExtractor = CurrentNetworkAttributesExtractor()

    @Test
    fun getNetworkAttributes_withCarrier() {
        val currentNetwork =
            CurrentNetwork(
                state = NetworkState.TRANSPORT_CELLULAR,
                subType = "aaa",
                carrier = Carrier(206, "ShadyTel", "usa", "omg", "US"),
            )

        val attributes = underTest.extract(currentNetwork).asMap()
        val expected =
            mapOf(
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CONNECTION_TYPE) to "cell",
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CONNECTION_SUBTYPE) to "aaa",
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CARRIER_NAME) to "ShadyTel",
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CARRIER_ICC) to "US",
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CARRIER_MCC) to "usa",
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CARRIER_MNC) to "omg",
            )
        assertThat(attributes).isEqualTo(expected)
    }

    @Test
    fun getNetworkAttributes_withoutCarrier() {
        val currentNetwork = CurrentNetwork(state = NetworkState.TRANSPORT_CELLULAR, subType = "aaa")

        val attributes = underTest.extract(currentNetwork).asMap()
        val expected =
            mapOf(
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CONNECTION_SUBTYPE) to "aaa",
                AttributeKey.stringKey(NetworkAttributes.NETWORK_CONNECTION_TYPE) to "cell",
            )
        assertThat(attributes).isEqualTo(expected)
    }
}
